### PR TITLE
Maintenance windows

### PR DESCRIPTION
**Pull request by mcalmer**
## What does this PR change?

Implementation of [Maintenance Windows RFC](https://github.com/uyuni-project/uyuni-rfc/blob/master/accepted/00073-maintenance-windows.md)

## GUI diff

No difference.

Before:

After:

- [ ] **DONE**

## Documentation
- No documentation needed: **add explanation. This can't be used if there is a GUI diff**
- Documentation issue was created: [Link for SUSE Manager contributors](https://github.com/SUSE/spacewalk/issues/new?template=ISSUE_TEMPLATE_DOCUMENTATION.md&labels=documentation&projects=SUSE/spacewalk/31), [Link for community contributors](https://github.com/uyuni-project/uyuni-docs/issues/new).
- (OPTIONAL) [Documentation PR](https://github.com/uyuni-project/uyuni-docs/pulls)

- [ ] **DONE**

## Test coverage
- No tests: **add explanation**
- Unit tests were added
- Cucumber tests were added

- [ ] **DONE**

## Links

Fixes #
Tracks # **add downstream PR, if any**

- [ ] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle" 
- [ ] Re-run test "java_pgsql_tests" (Test skipped, there are no changes to test)
- [ ] Re-run test "ruby_rubocop"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"     
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: java/code/src/com/suse/manager/maintenance/test/MaintenanceManagerTest.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        assertContains(names, "test server");
+
+        Optional<MaintenanceSchedule> dbScheduleOpt = mm.lookupMaintenanceScheduleByUserAndName(user, "test server");
+        assertNotNull(dbScheduleOpt.orElse(null));
+        MaintenanceSchedule dbSchedule = dbScheduleOpt.get();
+
+        assertEquals(user.getOrg(), dbSchedule.getOrg());
+        assertEquals("test server", dbSchedule.getName());
+        assertEquals(ScheduleType.SINGLE, dbSchedule.getScheduleType());
+        assertNull(dbSchedule.getCalendarOpt().orElse(null));
```
Comment (hustodemon) [inline java/code/src/com/suse/manager/maintenance/test/MaintenanceManagerTest.java:54]: General nitpick:
```suggestion
        assertTrue(dbSchedule.getCalendarOpt().isEmpty());
```

File: java/code/src/com/redhat/rhn/domain/action/ActionType.java [hunk 2]
```diff
@@ -30,6 +30,7 @@
     private String name;
     private Character triggersnapshot;
     private Character unlockedonly;
+    private boolean maintenancemodeonly;
```
[the remark this answers is not in the document]
Comment (mcalmer) [reply java/code/src/com/redhat/rhn/domain/action/ActionType.java:33]: ok

File: java/code/src/com/suse/manager/model/maintenance/MaintenanceCalendar.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "mtcal_seq")
+    @SequenceGenerator(name = "mtcal_seq", sequenceName = "suse_mtcal_id_seq", allocationSize = 1)
+    public Long getId() {
+        return id;
+    }
+
+    /**
+     * @return the organization
+     */
+    @ManyToOne(fetch = FetchType.LAZY)
```
[the remark this answers is not in the document]
Comment (mcalmer) [reply java/code/src/com/suse/manager/model/maintenance/MaintenanceCalendar.java:69]: ok

File: java/code/src/com/suse/manager/model/maintenance/MaintenanceSchedule.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "mtsched_seq")
+    @SequenceGenerator(name = "mtsched_seq", sequenceName = "suse_mtsched_id_seq", allocationSize = 1)
+    public Long getId() {
+        return id;
+    }
+
+    /**
+     * @return the organization
+     */
+    @ManyToOne(fetch = FetchType.LAZY)
```
[the remark this answers is not in the document]
Comment (mcalmer) [reply java/code/src/com/suse/manager/model/maintenance/MaintenanceSchedule.java:110]: ok

File: java/code/src/com/suse/manager/model/maintenance/MaintenanceSchedule.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    @JoinColumn(name = "ical_id", nullable = true)
+    protected MaintenanceCalendar getCalendar() {
+        return calendar;
+    }
+
+    /**
+     * @return the calendar as optional
+     */
+    @Transient
+    public Optional<MaintenanceCalendar> getCalendarOpt() {
```
[the remark this answers is not in the document]
Comment (mcalmer) [reply java/code/src/com/suse/manager/model/maintenance/MaintenanceSchedule.java:146]: That had other downsides if I remember. The behavior of hibernate is different when you assign it to getter or to fields. I better keep it like it is

File: java/code/src/com/redhat/rhn/common/util/http/HttpClientAdapter.java [hunk 6]
```diff
@@ -113,6 +115,9 @@ public HttpClientAdapter() {
 
         clientBuilder.setDefaultCredentialsProvider(credentialsProvider);
         Builder requestConfigBuilder = RequestConfig.custom()
+                .setConnectionRequestTimeout(5000)
+                .setConnectTimeout(Config.get().getInt(HTTP_CONNECTION_TIMEOUT, 5000))
+                .setSocketTimeout(Config.get().getInt(HTTP_SOCKET_TIMEOUT, 5000))
```
[the remark this answers is not in the document]
Comment (mcalmer) [reply java/code/src/com/redhat/rhn/common/util/http/HttpClientAdapter.java:120]: I think yes. Socket timeout means, 5 seconds without data going over the line.
But I think about changing the default to 10 seconds as this was a timeout used by salt-api for example.

File: java/code/src/com/suse/manager/maintenance/MaintenanceManager.java [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+     * List Maintenance Schedule Names belong to the given User
+     * @param user the user
+     * @return a list of Schedule names
+     */
+    public List<String> listScheduleNamesByUser(User user) {
+        @SuppressWarnings("unchecked")
+        Stream<String> names = getSession()
+            .createQuery("SELECT name FROM MaintenanceSchedule WHERE org = :org")
+            .setParameter("org", user.getOrg())
+            .stream();
```
[the remark this answers is not in the document]
Comment (mcalmer) [reply java/code/src/com/suse/manager/maintenance/MaintenanceManager.java:115]: ok

File: java/code/src/com/suse/manager/xmlrpc/maintenance/MaintenanceHandler.java [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+     * @xmlrpc.param #param_desc("string", "type", "Schedule type: single, multi")
+     * @xmlrpc.returntype
+     * #array_begin()
+     * $MaintenanceScheduleSerializer
+     * #array_end()
+     */
+    public MaintenanceSchedule createSchedule(User loggedInUser, String name, String type) {
+        ensureOrgAdmin(loggedInUser);
+        return mm.createMaintenanceSchedule(loggedInUser, name, ScheduleType.lookupByLabel(type),
+                Optional.empty());
```
[the remark this answers is not in the document]
Comment (mcalmer) [reply java/code/src/com/suse/manager/xmlrpc/maintenance/MaintenanceHandler.java:103]: ok

File: testsuite/features/init_clients/centos_ssh_minion.feature [hunk 9]
```diff
@@ -64,3 +64,7 @@ Feature: Bootstrap a SSH-managed CentOS minion and do some basic operations on i
   Scenario: Check events history for failures on SSH-managed CentOS minion
     Given I am on the Systems overview page of this "ceos_ssh_minion"
     Then I check for failed events on history event page
+
+@centos_minion
```
Comment (Bischoff) [inline testsuite/features/init_clients/centos_ssh_minion.feature:68]: * Please tag with `# WORKAROUND fix CentOS random boot failures`
* I suggest to move that commit to a separate PR if it works
* I asked Ricardo whether we should not do that at images level, rather

Comment (mcalmer) [reply testsuite/features/init_clients/centos_ssh_minion.feature]: It does not work, so I removed this commit.

Comment (Bischoff) [reply testsuite/features/init_clients/centos_ssh_minion.feature]: OK!

**Comment by juliogonzalez** [comment]
Dear developer:

A branch for SUSE Manager 4.1 is now created on the internal repository, and all package SPEC files in the master branch of this repository are now updated to be '4.2.0' version.

All developers, regardless of being SUSE employee or not, need to keep the following things in mind:

1. Please rebase this PR against `master`branch and fix any conflicts.
2. If this PR includes schema upgrades, change the directory for the migration scripts to the **correct one**. At this precise moment it is `schema/spacewalk/upgrade/susemanager-schema-4.2.0-to-susemanager-schema-4.2.1` but it could be a different one if this PR is still open after Uyuni 2020.08 is released.

SUSE employees must also keep in mind:

1. If you merge the PR here, you most likely want to backport it to the new `Manager-4.1` branch.
2. You will need to **adapt** the PR when backporting for at least the changelogs and schema migrations (if there are on your PR).

File: testsuite/features/step_definitions/common_steps.rb [hunk 10]
```diff
@@ -206,6 +206,7 @@
   steps %(
     When I am authorized as "admin" with password "admin"
     When I follow the left menu "Systems > Overview"
+    When I wait until I see "System Overview" text
```
Comment (Bischoff) [inline testsuite/features/step_definitions/common_steps.rb:209]: `And` please (and also on previous line)

**Pull request merged**